test/distributed/test_ddp_with_rpc.py

```diff
@@ -0,0 +1,541 @@
+    ]
+
+
+class TestDdpWithRpc(TestCase):
+    @classmethod
+    def _remote_worker_process(cls):
+        gLogger.info(f"Starting the remote worker...")
+        with RpcContext(
+            name=REMOTE_WORKER_NAME,
+            rank=REMOTE_WORKER_RANK,
+            world_size=WORLD_SIZE,
+            group_name="trainers/remote RPC",
+        ), ProcessGroupContext(
+            REMOTE_WORKER_NAME,
+            ranks=DDP_TRAINER_RANKS,
+            group_name="trainers_ddp",
+        ) as process_group:
+            gLogger.info(f"The remote worker is running.")
+
+        gLogger.info(f"Exiting remote worker.")
+        # exit to avoid run teardown() for fork processes
+        sys.exit(0)
+
+    def spawn_remote_worker(self):
+        remote_worker_process = multiprocessing.Process(
+            target=self._remote_worker_process, name=REMOTE_WORKER_NAME
+        )
+        remote_worker_process.start()
+        self.processes.append(remote_worker_process)
+
+    @classmethod
+    def _trainer_process(cls, rank: int):
+        gLogger.info(f"Starting the trainer #{rank}...")
+        with RpcContext(
+            name=TRAINER_NAME_TEMPLATE.format(rank),
+            rank=rank,
+            world_size=WORLD_SIZE,
+            group_name="trainers/remote RPC",
+        ):
+            gLogger.info(f"Trainer #{rank} is running.")
+
+        gLogger.info(f"Exiting trainer #{rank}...")
+        # exit to avoid run teardown() for fork processes
+        sys.exit(0)
+
+    def spawn_trainers(self):
+        for rank in range(1, NUM_TRAINERS + 1):
+            trainer = multiprocessing.Process(
+                target=self._trainer_process,
+                name=TRAINER_NAME_TEMPLATE.format(rank),
+                args=(rank,),
+            )
+            trainer.start()
+            self.processes.append(trainer)
+
+    @classmethod
+    def do_test_on_master(
+        cls,
+        ddp_mode: DdpMode,
+        trainer_method: Callable[[FeatureSet], None],
+        remote_em_rref: rpc.RRef,
+        remote_net_rref: rpc.RRef,
+    ):
+        trainer_rrefs = []
+        for rank, trainer in zip(DDP_TRAINER_RANKS, TRAINER_NAMES):
+            trainer_rrefs.append(
+                rpc.remote(
+                    trainer,
+                    Trainer,
+                    args=(remote_em_rref, remote_net_rref, ddp_mode, rank + 1),
+                )
+            )
+
+        training_examples = get_training_examples()
+        for epoch in range(NUM_EPOCH):
+            futures = []
+            for idx, trainer_rref in enumerate(trainer_rrefs):
+                futures.append(
+                    _remote_method_async(
+                        trainer_method, trainer_rref, training_examples[idx]
+                    )
+                )
+            log_loss = 0
+            for future in futures:
+                log_loss += future.wait()
+            gLogger.info(f"Log loss at epoch #{epoch}: {log_loss}")
```

Comment:
I see that we're trying to mimic what the training toolkit does here but its better to simplify the unit test using MultiProcessTestCase and have each process run something like the following:

```
def set_embedding_rref(emb):
  embedding_rref_ = emb

def test_ddp_with_rpc(self):
  if self.rank == 0:
    emb_rref = rpc.remote('worker1', create_emb, ...) // Assume embedding on rank 1
    // Share embedding with other ranks.
    for i in self.world_size:
      emb = rpc.sync(i, set_embedding_rref(emb_ref)
  
   dist.barrier() // Ensures all processes have reached this point and received the embedding rref.
   // Now just run a normal training loop which loops up the embedding table 
   emb_res = rpc.sync(1, emb_lookup, embedding_rref_)
   ......
```

Reply:
Reuse `MultiProcessTestCase` now.